File: finding-seams/kotlin/src/main/kotlin/findingseams/speakerfeedback/SpeakerFeedbackService.kt
package findingseams.speakerfeedback

data class EvaluateSpeakerResult(
    val status: String,
    val averageScore: Double?,
)

class SpeakerFeedbackService(
    private val sensor: FeedbackSensor = FeedbackSensor(),
) {
    private val alertThreshold = 2.2

    fun evaluateSpeaker(): EvaluateSpeakerResult {
        val scores = sensor.readScores()

        if (scores == null || scores.isEmpty()) {
            return EvaluateSpeakerResult("ALERT", null)
        }

        val validScores = scores.filter { it in 1..5 }
        if (validScores.isEmpty()) {
            return EvaluateSpeakerResult("ALERT", null)
        }

        val average = validScores.sum().toDouble() / validScores.size
        return if (average < alertThreshold) {
            EvaluateSpeakerResult("ALERT", average)
        } else {
            EvaluateSpeakerResult("OK", average)
        }
    }
}

File: finding-seams/csharp/FindingSeams/SpeakerFeedback/SpeakerFeedbackService.cs
namespace FindingSeams.SpeakerFeedback;

public record EvaluateSpeakerResult(string Status, double? AverageScore);

public class SpeakerFeedbackService
{
    private readonly FeedbackSensor _sensor = new();
    private readonly double _alertThreshold = 2.2;

    public EvaluateSpeakerResult EvaluateSpeaker()
    {
        var scores = _sensor.ReadScores();

        if (scores is null || scores.Count == 0)
            return new EvaluateSpeakerResult("ALERT", null);

        var validScores = scores.Where(s => s is >= 1 and <= 5).ToList();
        if (validScores.Count == 0)
            return new EvaluateSpeakerResult("ALERT", null);

        var average = validScores.Sum() / (double)validScores.Count;
        if (average < _alertThreshold)
            return new EvaluateSpeakerResult("ALERT", average);

        return new EvaluateSpeakerResult("OK", average);
    }
}
